### DISCUSSION TITLE
Ignoring 'file:' locations also ignores production Cordova app events

### DISCUSSION THREAD

**Discussion by dotmanish** [plausible/analytics#493] [Anything Else]
## Prerequisites
- [X] I have searched open and closed issues to make sure that the bug has not yet been reported.

## Bug report
**Describe the bug**
If plausible.js is loaded from within a Cordova app (HTML/Javascript), it checks that the location matches 'file:' and doesn't proceed. All Cordova based HTML pages have a 'file://' URL.

**Expected behavior**
Since using Plausible on screens within a Cordova app that is used across website and mobile apps is a valid use-case, this should be an argument that developers can pass on to plausible initialization to allow 'file://' local URLs.

**Comment by karahi** [comment]
Also facing this issue, cordova app doesn't track any events, presumably because of the file:// protocol. I'm using the https://www.npmjs.com/package/plausible-tracker package, not dropping the script in.

@JH7 on #445 seems to have posted a condition which would permit file:// urls, and plausible/analytics/tracker/src/plausible.js seems to use that condition to just emit a warning.

presuming that same condition is used in conjunction with the trackLocalhost init parameter, I'm not sure why events do get logged when using a localhost domain, but don't get logged with a file:// protocol. 

So...I'm confused.

edit: brainfart, the events are getting sent to plausible from the client, but its receiving 400 errors back…is this to do with the {Referrer,Origin, x-requested-with} request headers not matching the domain of the project?

**Reply by ukutaht** [reply]
> edit: brainfart, the events are getting sent to plausible from the client, but its receiving 400 errors back…is this to do with the {Referrer,Origin, x-requested-with} request headers not matching the domain of the project?

Unlikely. Can you post here the JSON body of the request that gives you a 400?

**Reply by karahi** [reply]
`{
  "n": "content_link_received",
  "u": "file:///android_asset/www/index.html",
  "d": "app.myohsoawesomeapp.com",
  "r": null,
  "w": 412,
  "h": 0
}`

It's not actually `app.myohsoawesomeapp.com`, its minimally different (but I've not not made it public yet). Anyway, I don't know whether those Object keys are supposed to be one character, or if something got chewed up by webpack since I'm using [the npm package](https://www.npmjs.com/package/plausible-tracker).

**Reply by ukutaht** [reply]
Thanks!

I gave it a try locally and it fails for the following reason: we deconstruct the URL of the event in our backend into `hostname` and `pathname`. In our database, `hostname` is required. When parsing `"file:///android_asset/www/index.html"`the hostname is `null` and so the database insert fails.

On localhost, the hostname gets parsed as `localhost` so it works OK.

I suppose we could be more liberal in what we accept. When the hostname is missing we could insert
```
pathname: "/android_asset/www/index.html",
hostname: "(none)"
```

Or something like this. What do you think?

**Reply by karahi** [reply]
Ahh, thanks for the response! Well that would indeed help for my issue.

But then a follow-up question might be: in what sense does the plausible.local.js script help hybrid apps? While they are served off the local host, they are not served from `localhost`.

Android apps using Cordova/PhoneGap/Ionic/capacitorjs (and I think the Webview component in react-native) all serve the webview content from a URL scheme like  `/android_asset/{{framework_specific_path}}`. 

Or `file:///var/mobile/Containers/Bundle/Application/{{uid}}/com.{{domain}}.{{appName}}.app/www/index.html` on iOS (apparently [1], and it seems to also be true for capacitorjs et al from perusing stackoverflow)


[1] https://medium.com/@barsh/cordova-navigating-to-local-content-from-a-website-3ed774c58d37

**Reply by riderx** [reply]
@karahi and @ukutaht in hybrid app we have serve location like this : 
<h4>Capacitor</a></h4>

Platform | Origin
-- | --
iOS | capacitor://localhost
Android | http://localhost

<h4 >Ionic WebView 3.x plugin on Cordova</a></h4>

Platform | Origin
-- | --
iOS | ionic://localhost
Android | http://localhost

<h4 >Ionic WebView 2.x plugin on Cordova</a></h4>

Platform | Origin
-- | --
iOS | http://localhost:8080
Android | http://localhost:8080

check it there : https://ionicframework.com/docs/troubleshooting/cors#capacitor

Current plausible code work for Ionic WebView 2.x and Android if i understand well

**Reply by ukutaht** [reply]
> But then a follow-up question might be: in what sense does the plausible.local.js script help hybrid apps? While they are served off the local host, they are not served from localhost.

Previously the script would not even attempt to send a pageview when running locally. Now it's sending it but depending on the URL scheme the backend may or may not accept it. Obviously this needs to be fixed on the backend but the `local.js` script variant is working as intended.

**Reply by riderx** [reply]
@ukutaht i would love to help on this too, but i just learn Elixir lang exist during searching in plausible code ^^

**Reply by ukutaht** [reply]
It's on my todo list but I can't give any promises as to when I can get to it

**Reply by ukutaht** [reply]
The fix for this is now live: https://github.com/plausible/analytics/pull/1335

Thanks @RobertJoonas !

**Reply by riderx** [reply]
Thank's a lot @RobertJoonas

**Comment by metmarkosaric** [comment]
We now have the plausible.local.js extension that allows you to track stats on localhost too. [Read more here](https://plausible.io/docs/script-extensions). This should help with Cordova and similar hybrid apps too. Thanks @riderx for this contribution!

**Reply by karahi** [reply]
Will this get packaged up into [the npm package](https://www.npmjs.com/package/plausible-tracker)? Seems a bit wasteful to eat e.g. a Cloudflare Workers call-quota (or whatever) to serve a teeny tiny script, especially as most hybrid apps are bundled anyway.

**Reply by riderx** [reply]
@karahi in the npm package it seems the option `trackLocalhost` exist : https://github.com/plausible/plausible-tracker/blob/a5f8e946263b4e9898c729293f74a095ef71cfcb/src/lib/tracker.ts#L220
